### PR TITLE
wgengine/filter: implement a destination IP pre-filter.

### PR DESCRIPTION
**Pull request by danderson**
This change is [Reviewable](https://reviewable.io/reviews/tailscale/tailscale/398)

### CODE REVIEW

File: ipn/local.go [hunk 1]
```diff
@@ -391,26 +391,35 @@ func (b *LocalBackend) Start(opts Options) error {
...
+	b.mu.Lock()
+	advRoutes := b.prefs.AdvertiseRoutes
+	b.mu.Unlock()
+	localNets := wgCIDRsToFilter(netMap.Addresses, advRoutes)
+
+	if b.shieldsAreUp() {
 		// Shields up, block everything
 		b.logf("netmap packet filter: (shields up)")
-		b.e.SetFilter(filter.NewAllowNone(b.logf))
+		b.e.SetFilter(filter.New(nil, localNets, nil, b.logf))
```
Comment (bradfitz) [inline ipn/local.go:409]: whenever I see a call site like this, I start rethinking APIs nil false 0 0 true nil

Comment (bradfitz) [reply ipn/local.go:409]: At minimum, put the nils in local variables so it's obvious what is nil.

Or `nil /* foo */, localNets, nil /* bar */` but that looks so foreign in Go.

File: wgengine/filter/filter.go [hunk 2]
```diff
@@ -23,9 +23,10 @@ type filterState struct {
 
 // Filter is a stateful packet filter.
 type Filter struct {
-	logf    logger.Logf
-	matches Matches
-	state   *filterState
+	logf      logger.Logf
+	localNets []Net
```
Comment (bradfitz) [inline wgengine/filter/filter.go:27]: document this field? (or all of them if you're feeling fancy)

File: wgengine/filter/filter.go [hunk 3]
```diff
@@ -75,21 +76,22 @@ var MatchAllowAll = Matches{
...
-// New creates a new packet Filter with the given Matches rules.
-// If shareStateWith is non-nil, the returned filter shares state
-// with the previous one, to enable rules to be changed at runtime
-// without breaking existing flows.
-func New(matches Matches, shareStateWith *Filter, logf logger.Logf) *Filter {
+// New creates a new packet Filter with the given Matches rules and
+// known local destination networks. If shareStateWith is non-nil, the
+// returned filter shares state with the previous one, to enable rules
+// to be changed at runtime without breaking existing flows.
+func New(matches Matches, localNets []Net, shareStateWith *Filter, logf logger.Logf) *Filter {
```
Comment (bradfitz) [inline wgengine/filter/filter.go:94]: how about just export the Filter fields and let callers write a composite literal with the ones they care about?

A popular Go style for things that require or want constructors is have a constructor for the required fields, return a *T and let caller set exported optional fields, then start using the thing. Document "must set these fields before use and not after", etc.

File: ipn/local.go [hunk 4]
```diff
@@ -788,6 +797,21 @@ func routerConfig(cfg *wgcfg.Config, prefs *Prefs, dnsDomains []string) *router.
 	return rs
 }
 
+func wgCIDRsToFilter(cidrs ...[]wgcfg.CIDR) (ret []filter.Net) {
```
Comment (bradfitz) [inline ipn/local.go:800]: document this? (the `...` along with the `[]` caught me off guard at first. but it's fine.)

Comment (bradfitz) [reply ipn/local.go:800]: Maybe call it `cidrSets`, though? It's double plural, really.

File: wgengine/filter/filter.go [hunk 5]
```diff
@@ -75,21 +76,22 @@ var MatchAllowAll = Matches{
...
+	return New(nil, nil, nil, logf)
 }
 
-// New creates a new packet Filter with the given Matches rules.
-// If shareStateWith is non-nil, the returned filter shares state
-// with the previous one, to enable rules to be changed at runtime
-// without breaking existing flows.
-func New(matches Matches, shareStateWith *Filter, logf logger.Logf) *Filter {
+// New creates a new packet Filter with the given Matches rules and
+// known local destination networks. If shareStateWith is non-nil, the
```
Comment (bradfitz) [inline wgengine/filter/filter.go:91]: but what does it do with these known local destination networks? accept, deny, require them to be in it? what does nil mean?

File: ipn/local.go (hunk 1 above)
Comment (danderson) [reply ipn/local.go:409]: Agree, this API is getting a little unwieldy. For now I'll define local vars to make the call site easier to read.

**Comment by danderson** [comment]
Was pushed separate from this PR.

**Pull request closed without merging**